bundles/org.openhab.binding.paradoxalarm/src/main/java/org/openhab/binding/paradoxalarm/internal/communication/crypto/EncryptionHandler.java

```diff
@@ -0,0 +1,307 @@
+        }
+
+        printArray("Decrypted array", result);
+        return result;
+    }
+
+    private void printArray(String description, byte[] array) {
+        ParadoxUtil.printByteArray(description, array, array.length);
+    }
+
+    private byte[] fillArray(byte[] keyBytes) {
+        byte[] byteArray = new byte[keyBytes.length];
+        for (int i = 0; i < keyBytes.length; i++) {
+            byteArray[i] = (byte) (keyBytes[i] & 0xFF);
+        }
+
+        byte[] expandedArray = ParadoxUtil.extendArray(byteArray, KEY_ARRAY_LENGTH);
+        return expandedArray;
+    }
+
+    public void updateKey(byte[] newKey) {
+        expandedKey = new int[KEY_LENGTH];
+        expandKey(newKey);
+    }
+
+    private void expandKey(byte[] input) {
+        // fill array to 32th byte with 0xEE
+        byte[] filledArray = fillArray(input);
+
+        int[] temp = { 0, 0, 0, 0 };
+        for (int i = 0; i < 4; i++) {
+            for (int j = 0; j < 4; j++) {
+                expandedKey[j * 4 + i] = filledArray[i * 4 + j] & 0xFF;
+            }
+            for (int j = 0; j < 4; j++) {
+                expandedKey[j * 4 + i + 16] = filledArray[i * 4 + j + 16] & 0xFF;
+            }
+        }
+
+        final int[] s = EncryptionHandlerConstants.S;
+        for (int i = 8; i < 60; i++) {
+            for (int j = 0; j < 4; j++) {
+                temp[j] = expandedKey[(((i - 1) & 0xfc) << 2) + ((i - 1) & 0x03) + j * 4];
+            }
+
+            if (i % 8 == 0) {
+                for (int j = 0; j < 4; j++) {
+                    temp[j] = s[temp[j]];
+                }
+                int tmp = temp[0];
+
+                for (int j = 1; j < 4; j++) {
+                    temp[j - 1] = temp[j];
+                }
+
+                temp[3] = tmp;
+                temp[0] ^= EncryptionHandlerConstants.RCON[(i / 8 - 1)];
+            } else if (i % 8 == 4) {
+                for (int j = 0; j < 4; j++) {
+                    temp[j] = s[temp[j]];
+                }
+            }
+
+            for (int j = 0; j < 4; j++) {
+                expandedKey[((i & 0xfc) << 2) + (i & 0x03)
+                        + j * 4] = expandedKey[(((i - 8) & 0xfc) << 2) + ((i - 8) & 0x03) + j * 4] ^ temp[j];
+            }
+        }
+    }
+
+    private int gmul(int c, int b) {
+        int s;
+        int q;
+        int z = 0;
+        s = lTable[c] + lTable[b];
+        s %= 255;
+        /* Get the antilog */
+        s = aTable[s];
+        /*
+         * Now, we have some fancy code that returns 0 if either
+         * a or b are zero; we write the code this way so that the
+         * code will (hopefully) run at a constant speed in order to
+         * minimize the risk of timing attacks
+         */
+        q = s;
+        if (c == 0) {
+            s = z;
+        } else {
+            s = q;
+        }
+        if (b == 0) {
+            s = z;
+        } else {
+            q = z;
+        }
```

Comment:
The final assignment `q = z` serves no useful purpose here since the method returns the next line. This code is wrong. I suspect this code is somehow not arranged correctly.
```suggestion
        if (b == 0) {
            s = z;
        } else {
            q = z;
        }
        if (c == 0) {
            s = z;
        } else {
            s = q;
        }
```

That said, this is very convoluted for something that can be written much simpler.
```suggestion
        if (b == 0 | c == 0) {
            return 0;
        }
```

Reply:
Absolutely. Refactored that in a simmilar to the suggested way. Thanks !